### PR TITLE
BUGFIX: 4742 - document uri path projection adjustments

### LINKED ISSUE #4742: BUG: 9.0 - Translating document nodes leads to wrong ordering
Currently the ordering behavior of the CR is undefined for the following scenarios:

- [x] CreateNodeAggregateWithNode with fallbacks
- [x] Create Peer Variant
- [x] Create Generalization

This leads to the problem that different projections implement ordering differently then; leading to subtle bugs.

For instance, the `DocumentUriPathProjection` needs the ordering as well to find the first child node for FirstChild-shortcuts -- and it currently does it **differently** than the Content Graph.

That is what I originally wanted to fix; when I stumbled into this rabbit hole ;)

### Current Behavior

On the demo site, let's say you have the following english nodes:

- a
- b
- c
- d

=> now, you translate `b` to german; leading to the following structure:

- b

=> now, you translate `a` to german, leading to the following structure:

- b
- a

=> this is wrong, expected would be:

```
- a
- b
```

### Fixing Plan

- [x] NodePeerVariantWasCreated
  - **highest priority**, because it is relevant when simply having a language dimension and starting to translate
  - [x] extend event to have s.th. like `PositionMapping` (modeled after `MoveNodeMapping`)
  - [x] in command handler, use same logic to determine positionMapping as in MoveNode
  - [x] write testcase
  - [x] ensure ordering is respected in graph projection
  - [x] ensure ordering is respected in Routing projection
- [x] NodeGeneralizationVariantWasCreated
- [x] NodeAggregateWithNodeWasCreated with fallbacks

NOTE: NodeSpecializationVariantWasCreated does not need to be adjusted, because it only leads to re-wiring of edges, and not to new edges.

### PR DESCRIPTION
**Pull request by nezaniel**
Resolves: https://github.com/neos/neos-development-collection/issues/4742

Makes use of the new InterdimensionalSiblings model to fix the uri path projection regarding sibling positioning.
The respective edge case tests have been taken over from https://github.com/neos/neos-development-collection/pull/4961, wich this is based upon

**Upgrade instructions**

none

**Review instructions**

DocumentUriPathProjection::connectNodeWithSiblings seemed a bit weird since it considered the provided node as its own preceding sibling, but this has to be this way for node move, while it must not be for node variation. The respective switch is implemented, making it, well, not less weird. But the tests pass, so...

**Checklist**

- [x] Code follows the PSR-12 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the 9.0 Branch
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php [hunk 1]
```diff
@@ -908,10 +909,14 @@ private function disconnectNodeFromSiblings(DocumentNodeInfo $node): void
...
 
+    /**
+     * @param bool $includingSelf whether the node itself is to be considered its own sibling. true for moving, false for variation
+     */
     private function connectNodeWithSiblings(
         DocumentNodeInfo $node,
         NodeAggregateId $parentNodeAggregateId,
-        ?NodeAggregateId $newSucceedingNodeAggregateId
+        ?NodeAggregateId $newSucceedingNodeAggregateId,
+        bool $includingSelf,
```
Comment (mhsdesign) [inline Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php]: ugly alert

File: Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php [hunk 2]
```diff
@@ -657,7 +658,7 @@ private function moveNode(
     ): void {
         $this->disconnectNodeFromSiblings($node);
 
-        $this->connectNodeWithSiblings($node, $newParentNodeAggregateId, $newSucceedingNodeAggregateId);
+        $this->connectNodeWithSiblings($node, $newParentNodeAggregateId, $newSucceedingNodeAggregateId, true);
```
Comment (bwaidelich) [inline Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php]: @nezaniel I can change this to 
```suggestion
        $this->connectNodeWithSiblings($node, $newParentNodeAggregateId, $newSucceedingNodeAggregateId, false);
```
without any tests failing..
That makes me wonder whether we could just remove the `$includingSelf` parameter below

Comment (nezaniel) [reply Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php]: mmh, weird... the projection looked really funny (and not in a good way) when I omitted this, will re-check and add failing tests

Comment (nezaniel) [reply Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php]: I withdraw my point and claim the exact opposite: You're correct, seems like I fixed my weird stuff along the way by accident... 

I removed the switch, looks a lot better now.

**Review by bwaidelich — APPROVED** [review]
The best of all outcomes \o/

File: Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php (hunk 1 above)
Comment (nezaniel) [reply Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php]: /fixed

File: Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathFinder.php [hunk 3]
```diff
@@ -195,6 +195,28 @@ public function getLastChildNode(
         );
     }
 
+    /**
+     * @throws NodeNotFoundException
+     * @internal
+     */
+    public function getLastChildNodeNotBeing(
```
Comment (mhsdesign) [inline Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathFinder.php:202]: naming was a little hard to grasp but when looking at the parameter `$excludedNodeAggregateId` it _should_ explain.

File: Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php [hunk 4]
```diff
@@ -927,24 +928,34 @@ private function connectNodeWithSiblings(
                 ['precedingNodeAggregateId' => $node->getNodeAggregateId()->value]
             );
         } else {
-            $newPrecedingNode = $this->tryGetNode(fn () => $this->getState()->getLastChildNode(
+            $newPrecedingNode = $this->tryGetNode(fn () => $this->getState()->getLastChildNodeNotBeing(
                 $parentNodeAggregateId,
-                $node->getDimensionSpacePointHash()
+                $node->getDimensionSpacePointHash(),
+                $node->getNodeAggregateId()
```
Comment (mhsdesign) [inline Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php:934]: maybe like this?

```suggestion
                excludedNodeAggregateId: $node->getNodeAggregateId()
```

**Review by mhsdesign — COMMENTED** [review]
Looks good by reading (kindof) ... only +0.5 as i havent actually tried this out yet.
but feel free to merge already ;)
